src/engine/SoftmaxConstraint.cpp

```diff
@@ -0,0 +1,291 @@
+            return;
+        setLowerBound( variable, bound );
+    }
+    // TODO: bound derivation
+}
+
+void SoftmaxConstraint::notifyUpperBound( unsigned variable, double bound )
+{
+    ASSERT( participatingVariable( variable ) );
+    if ( _statistics )
+        _statistics->incLongAttribute
+            ( Statistics::NUM_BOUND_NOTIFICATIONS_TO_TRANSCENDENTAL_CONSTRAINTS );
+
+    if ( _boundManager == nullptr )
+    {
+        if ( existsUpperBound( variable ) &&
+             !FloatUtils::lt( bound, getUpperBound( variable ) ) )
+            return;
+        setUpperBound( variable, bound );
+    }
+    // TODO: bound derivation
+}
+
+bool SoftmaxConstraint::participatingVariable( unsigned variable ) const
+{
+    return ( _inputs.exists( variable ) || _outputs.exists( variable ) );
+}
+
+List<unsigned> SoftmaxConstraint::getParticipatingVariables() const
+{
+    List<unsigned> toReturn;
+    for ( const auto &var : _inputs )
+        toReturn.append( var );
+    for ( const auto &var : _outputs )
+        toReturn.append( var );
+    return toReturn;
+}
+
+void SoftmaxConstraint::updateVariableIndex( unsigned oldIndex, unsigned newIndex )
+{
+    ASSERT( participatingVariable( oldIndex ) );
+    ASSERT( !participatingVariable( newIndex ) );
+
+    if ( _assignment.exists( oldIndex ) )
+    {
+        _assignment[newIndex] = _assignment.get( oldIndex );
+        _assignment.erase( oldIndex );
+    }
+
+    if ( _lowerBounds.exists( oldIndex ) )
+    {
+        _lowerBounds[newIndex] = _lowerBounds.get( oldIndex );
+        _lowerBounds.erase( oldIndex );
+    }
+
+    if ( _upperBounds.exists( oldIndex ) )
+    {
+        _upperBounds[newIndex] = _upperBounds.get( oldIndex );
+        _upperBounds.erase( oldIndex );
+    }
+
+    for ( unsigned i = 0; i < _inputs.size(); ++i )
+    {
+        if ( _inputs[i] == oldIndex )
+            _inputs[i] = newIndex;
+        if ( _outputs[i] == oldIndex )
+            _outputs[i] = newIndex;
+    }
+}
+
+void SoftmaxConstraint::eliminateVariable( __attribute__((unused)) unsigned variable,
+                                        __attribute__((unused)) double fixedValue )
+{
+    throw MarabouError( MarabouError::FEATURE_NOT_YET_SUPPORTED,
+                        "Eliminate variable from a Softmax constraint" );
+}
+
+bool SoftmaxConstraint::constraintObsolete() const
+{
+    return _inputs.size() == 0;
+}
+
+void SoftmaxConstraint::getEntailedTightenings( List<Tightening> &tightenings ) const
+{
+    for ( unsigned i = 0; i < _outputs.size(); ++i )
+    {
+        tightenings.append( Tightening( _outputs[i], 0, Tightening::LB ) );
+        tightenings.append( Tightening( _outputs[i], 1, Tightening::UB ) );
+
+        // First get tighter lower bound
+        bool validBound = true;
+        Vector<double> values;
+        for ( unsigned j = 0; j < _inputs.size(); ++j )
+        {
+            if ( i == j )
+            {
+                if ( existsLowerBound( _inputs[j] ) &&
```

Comment:
I didn't understand what was going on here. Are we just taking the existing bounds and propagating them as is?